c_src/erl_optics.c

```diff
@@ -540,18 +538,27 @@ static void backend_free_eo(void *ctx){
 static ERL_NIF_TERM eo_optics_poll(
     ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
 {
-    struct optics *optics = get_optics(env, argv[0]);
-    if (!optics) return ERROR("get_optics");
-    if (!poller) return ERROR("poller_allocation_error");
+    if (enif_rwlock_tryrwlock(poll_lock)) return ERROR("concurrent_polling");
 
+    struct optics *optics = get_optics(env, argv[0]);
+    if (!optics) {
+        enif_rwlock_rwunlock(poll_lock);
+        return ERROR("get_optics");
+    }
+    if (!poller) {
+        enif_rwlock_rwunlock(poll_lock);
+        return ERROR("poller_allocation_error");
```

Comment:
The problem with having multiple unlock statement is that it's somewhat error prone. As an example, if someone adds an extra check or branch that can error out then that person also has to remember to properly call the unlock function.

Luckily, for the first checks it's pretty easy to deal with by just grabbing the lock after you make the calls. For the later checks it's a bit trickier but this usually how I structure the code for stuff like this:
```
ERL_NIF_TERM ret = {0};
{
  if (enif_rwlock_tryrwlock(poll_lock)) return ERROR("concurrent_polling");

  if (erlang_poller) { ctx->... = ...; }
  if (!optics_poller_poll(poller)) ret = ERROR("poller_something_something");

  enif_rwlock_rwunlock(poll_lock);
}

if (ret) return ret;
return enif_make_tuple2(env, atom_ok, erlang_poller ? atom_ok : ctx->map);
```
Note that the extra scoping does nothing but I find that it acts as a strong indicator of the lifetime of the lock and a reminder to not just randomly return from within that scope. Feel free to adapt to whoever you see fit (I'm not entirely sure that everything I wrote there actually compiles).

Reply:
I like that. Done.